feat: EPI-1325: Create 'Dispensed medications' table

khoi-bes opened the pull request:
### Changes

_Add a brief description of the changes in this PR to help give the reviewer context._

### Deploys

- [ ] **Deploy to Tamanu Internal** <!-- #deploy -->

### Tests

- [ ] **Run E2E Tests** <!-- #e2e -->

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

<!-- Thank you! -->

namhoangbes committed:
feat: EPI-1323: implement apis for medications dispensing flow

namhoangbes committed:
feat: EPI-1323: implement apis for medications dispensing flow

huytranbes committed:
Merge branch 'epic-medications-dispensing' into feat/EPI-1323-apis-for-dispensing-medications-flow

huytranbes committed:
feat: EPI-1356: Add display ID to pharmacy order prescriptions and implement dispensing workflow modal

- Introduced a new column 'display_id' in the 'pharmacy_order_prescriptions' table to uniquely identify prescriptions.
- Updated the PharmacyOrderPrescription model to include 'displayId' with a default value generated by a utility function.
- Enhanced the medication retrieval API to include 'displayId' in the response.
- Created a new DispenseMedicationWorkflowModal component for managing medication dispensing, including validation and printing of labels.
- Added MedicationLabelPrintModal and MedicationLabel components for rendering and printing medication labels.
- Updated MedicationRequestsTable to integrate the new dispensing workflow and display relevant information.

huytranbes committed:
Merge branch 'feat/EPI-1323-apis-for-dispensing-medications-flow' of https://github.com/beyondessential/tamanu into feat/EPI-1323-dispense-medications-workflow

huytranbes committed:
feat: Add human-readable display ID to pharmacy order prescriptions documentation

- Introduced documentation for a new 'display_id' field in the pharmacy order prescriptions model, providing a human-readable request number for prescription orders.
- Updated the YAML configuration to include 'display_id' with appropriate data type and description.

huytranbes committed:
feat: Update MedicationLabel component to use dynamic dimensions

- Refactored the MedicationLabel component to accept dynamic width and height based on user settings.
- Integrated useSettings hook to retrieve label dimensions, enhancing flexibility for different printing requirements.

huytranbes committed:
chore: Remove DispenseMedicationWorkflowModal test file

- Deleted the test file for DispenseMedicationWorkflowModal as part of the cleanup process.
- This file contained tests for helper functions related to medication dispensing workflow.

huytranbes committed:
feat: Enhance DispenseMedicationWorkflowModal with validation and input handling

- Added validation schema using Yup for dispensing workflow, ensuring required fields and conditions are met.
- Refactored input handling for quantity and instructions fields to improve user experience and error management.
- Updated MedicationLabel component to display quantity with dynamic unit labels for better clarity.
- Removed unnecessary data tests from pharmacy order prescriptions documentation.

huytranbes committed:
feat: Add inline validation messages to DispenseMedicationWorkflowModal

- Implemented helper text for quantity and instructions fields to display validation errors when required fields are not filled.
- Enhanced user experience by providing immediate feedback on input errors during the medication dispensing workflow.

huytranbes committed:
refactor: Simplify input handling and enhance MedicationLabel component

- Refactored the InstructionsInput and QuantityInput components to streamline the change handling logic.
- Updated the MedicationLabel component to dynamically display quantity with units, improving clarity for users.
- Added support for units in the MedicationLabel prop types to ensure accurate label rendering.

huytranbes committed:
feat: Enhance DispenseMedicationWorkflowModal with detailed error handling

- Introduced item-specific error tracking for quantity and instructions fields in the DispenseMedicationWorkflowModal.
- Updated validation logic to provide immediate feedback on input errors, improving user experience during medication dispensing.
- Refactored input components to utilize the new error state, ensuring accurate display of validation messages.

huytranbes committed:
refactor: Streamline input handling and validation in DispenseMedicationWorkflowModal

- Simplified the change handling logic in InstructionsInput and QuantityInput components by removing unnecessary useCallback hooks.
- Enhanced the validation logic for selected items and stock status by eliminating useMemo hooks where not needed, improving performance.
- Refactored handleClose, handleSelectAll, handleSelectRow, handleQuantityChange, and handleInstructionsChange functions for better readability and maintainability.

huytranbes committed:
refactor: Improve prescription data handling in DispenseMedicationWorkflowModal

- Streamlined the extraction of prescription properties by destructuring in the buildInstructionText function, enhancing readability.
- Simplified item selection and quantity handling by using destructuring in the handleSelectAll and handleQuantityChange functions.
- Updated the mapping of dispensable items to improve clarity and maintainability, ensuring consistent handling of prescription data throughout the component.

huytranbes committed:
fix: Update display_id data type in pharmacy_order_prescriptions model

- Changed the data type of the display_id field from character varying to character varying(255) to ensure proper handling of longer display IDs.
- This adjustment enhances the model's capability to accommodate varying lengths of display IDs in prescription orders.

huytranbes committed:
update buildInstructionText

namhoangbes committed:
fix: fix fetch remaining repeats of medications

huytranbes committed:
feat: Add discharge prescription handling in medication workflow

- Enhanced the medication retrieval process to include the isDischargePrescription property, allowing for better management of discharge prescriptions.
- Updated the DispenseMedicationWorkflowModal to disable items that are discharge prescriptions with no remaining repeats, improving user experience during medication dispensing.

huytranbes committed:
Merge branch 'feat/EPI-1323-apis-for-dispensing-medications-flow' into feat/EPI-1323-dispense-medications-workflow

khoi-bes committed:
feat: EPI-1325: Create 'Dispensed medications' table

huytranbes commented on packages/web/app/components/MedicationDispensesTable.jsx:
```diff
@@ -0,0 +1,217 @@
+        <NoDataContainer>
+          <TranslatedText
+            stringId="medication-dispenses.list.noData"
+            fallback="No active medication requests to display."
```
Should be "No dispensed medications to display."

khoi-bes committed:
addressed feedbacks

namhoangbes committed:
fix: EPI-1323: fix api get list dispensable medications

khoi-bes committed:
Merge branch 'epic-medications-dispensing' into feat/EPI-1323-apis-for-dispensing-medications-flow

khoi-bes committed:
refactor(medication): Streamline dispensable medications API and dispense logic

* Refactored the dispensable medications endpoint to improve query structure and include stock status.
* Consolidated dispense logic to enhance transaction handling and error management.
* Removed outdated index from medication dispenses table migration.

khoi-bes committed:
fix(medication): Remove 'createdAt' attribute from pharmacy order prescriptions query

* Updated the pharmacy order prescriptions query to exclude the 'createdAt' attribute for improved data handling.

khoi-bes committed:
refactor(medication): Simplify prescription deletion logic in dispense process

* Removed unnecessary transaction parameter from pharmacy order prescription queries.
* Streamlined the logic for identifying ineligible pharmacy order prescriptions by adjusting the remaining repeats check.

khoi-bes committed:
fix(medication): Adjust remaining repeats calculation and validate pharmacy order consistency

* Updated the calculation of remaining repeats in PharmacyOrderPrescription to account for the first dispense.
* Added validation to ensure all prescriptions belong to the same pharmacy order during the dispense process.
* Modified the dispensable medications query schema to use the correct UUID type for patientId.

huytranbes committed:
Merge branch 'feat/EPI-1323-apis-for-dispensing-medications-flow' of https://github.com/beyondessential/tamanu into feat/EPI-1323-dispense-medications-workflow

khoi-bes committed:
Merge remote-tracking branch 'origin/epic-medications-dispensing' into feat/EPI-1325-create-dispensed-medications-table

khoi-bes committed:
refactor(medication): Remove dispensable medications endpoint and streamline dispense logic

- Eliminated the dispensable medications API endpoint to simplify the medication handling process.
- Retained the dispense logic while ensuring it remains functional and efficient.
- Updated related schemas to reflect the removal of the dispensable medications functionality.

khoi-bes committed:
refactor(medication): Update stock status constants in DispenseMedicationWorkflowModal

- Replaced STOCK_STATUS_LABELS and STOCK_STATUSES with DRUG_STOCK_STATUS_LABELS and DRUG_STOCK_STATUSES for improved clarity and consistency.
- Adjusted the getStockStatus function and related components to utilize the new constants.

khoi-bes committed:
Merge branch 'epic-medications-dispensing' into feat/EPI-1325-create-dispensed-medications-table

khoi-bes committed:
Merge branch 'epic-medications-dispensing' into feat/EPI-1325-create-dispensed-medications-table

khoi-bes committed:
Merge branch 'feat/EPI-1325-create-dispensed-medications-table' of bes:beyondessential/tamanu into feat/EPI-1325-create-dispensed-medications-table

huytranbes committed:
fix(medication): Update stock status constants for drug dispensing workflow

huytranbes committed:
feat(medication): Enhance stock status handling in medication components

huytranbes committed:
feat(medication): Add onDispenseSuccess callback to refresh medication requests table

khoi-bes committed:
Merge branch 'epic-medications-dispensing' into feat/EPI-1325-create-dispensed-medications-table

khoi-bes committed:
Merge branch 'epic-medications-dispensing' into feat/EPI-1323-dispense-medications-workflow

huytranbes committed:
feat(medication): Update medication attributes and add facility query for dispensing workflow

huytranbes committed:
refactor(medication): Clean up and optimize DispenseMedicationWorkflowModal component

khoi-bes committed:
fix(medication): Correct remaining repeats calculation and enforce patient consistency in prescriptions

khoi-bes committed:
Merge branch 'feat/EPI-1323-dispense-medications-workflow' of bes:beyondessential/tamanu into feat/EPI-1323-dispense-medications-workflow

khoi-bes committed:
Merge branch 'feat/EPI-1323-dispense-medications-workflow' into feat/EPI-1325-create-dispensed-medications-table

khoi-bes committed:
Merge branch 'feat/EPI-1325-create-dispensed-medications-table' of bes:beyondessential/tamanu into feat/EPI-1325-create-dispensed-medications-table

khoi-bes committed:
Merge branch 'epic-medications-dispensing' into feat/EPI-1325-create-dispensed-medications-table

khoi-bes committed:
refactor: Remove minifyAliases flag from database connection options and enhance patient data retrieval in medication API response
